File: attocode/src/shared/persistence.ts
/**
 * Unified Persistence Adapters
 *
 * Thin interface over different storage backends (JSON files, SQLite).
 * Provides a common API for save/load/list/delete operations.
 *
 * Existing stores (SQLiteStore, SwarmStateStore) continue unchanged.
 * These adapters are opt-in wrappers for unified access patterns.
 */

import fs from 'node:fs/promises';
import path from 'node:path';
import Database from 'better-sqlite3';

// =============================================================================
// MAP SERIALIZATION (reused from swarm-state-store pattern)
// =============================================================================

/** JSON replacer that serializes Maps as { __type: 'Map', entries: [...] } */
function mapReplacer(_key: string, value: unknown): unknown {
  if (value instanceof Map) {
    return { __type: 'Map', entries: [...value.entries()] };
  }
  return value;
}

/** JSON reviver that deserializes Maps from { __type: 'Map', entries: [...] } */
function mapReviver(_key: string, value: unknown): unknown {
  if (
    value &&
    typeof value === 'object' &&
    (value as Record<string, unknown>).__type === 'Map'
  ) {
    return new Map((value as { entries: [unknown, unknown][] }).entries);
  }
  return value;
}

// =============================================================================
// INTERFACE
// =============================================================================

export interface PersistenceAdapter {
  /** Save data under a key within a namespace */
  save(namespace: string, key: string, data: unknown): Promise<void>;
  /** Load data by namespace + key */
  load(namespace: string, key: string): Promise<unknown | null>;
  /** List all keys in a namespace */
  list(namespace: string): Promise<string[]>;
  /** Delete a key */
  delete(namespace: string, key: string): Promise<boolean>;
  /** Check if a key exists */
  exists(namespace: string, key: string): Promise<boolean>;
}

// =============================================================================
// JSON FILE ADAPTER
// =============================================================================

/**
 * JSON file-based persistence.
 * Each namespace is a directory, each key is a JSON file.
 */
export class JSONFilePersistenceAdapter implements PersistenceAdapter {
  constructor(private baseDir: string) {}

  private keyPath(namespace: string, key: string): string {
    return path.join(this.baseDir, namespace, `${key}.json`);
  }

  private nsDir(namespace: string): string {
    return path.join(this.baseDir, namespace);
  }

  async save(namespace: string, key: string, data: unknown): Promise<void> {
    const filePath = this.keyPath(namespace, key);
    await fs.mkdir(path.dirname(filePath), { recursive: true });
    await fs.writeFile(filePath, JSON.stringify(data, mapReplacer, 2));
  }

  async load(namespace: string, key: string): Promise<unknown | null> {
    try {
      const content = await fs.readFile(this.keyPath(namespace, key), 'utf-8');
      return JSON.parse(content, mapReviver) as unknown;
    } catch {
      return null;
    }
  }

  async list(namespace: string): Promise<string[]> {
    try {
      const entries = await fs.readdir(this.nsDir(namespace));
      return entries
        .filter((e) => e.endsWith('.json'))
        .map((e) => e.slice(0, -5));
    } catch {
      return [];
    }
  }

  async delete(namespace: string, key: string): Promise<boolean> {
    try {
      await fs.unlink(this.keyPath(namespace, key));
      return true;
    } catch {
      return false;
    }
  }

  async exists(namespace: string, key: string): Promise<boolean> {
    try {
      await fs.access(this.keyPath(namespace, key));
      return true;
    } catch {
      return false;
    }
  }
}

// =============================================================================
// SQLITE ADAPTER
// =============================================================================

const CREATE_TABLE_SQL = `
  CREATE TABLE IF NOT EXISTS kv_store (
    namespace TEXT NOT NULL,
    key TEXT NOT NULL,
    data TEXT NOT NULL,
    updated_at INTEGER NOT NULL,
    PRIMARY KEY (namespace, key)
  )
`;

/**
 * SQLite-based persistence.
 * Uses a generic key-value table with namespace partitioning.
 */
export class SQLitePersistenceAdapter implements PersistenceAdapter {
  private db: InstanceType<typeof Database>;

  constructor(dbPath: string) {
    this.db = new Database(dbPath);
    this.db.pragma('journal_mode = WAL');
    this.db.exec(CREATE_TABLE_SQL);
  }

  async save(namespace: string, key: string, data: unknown): Promise<void> {
    this.db
      .prepare(
        `INSERT OR REPLACE INTO kv_store (namespace, key, data, updated_at)
         VALUES (?, ?, ?, ?)`,
      )
      .run(namespace, key, JSON.stringify(data, mapReplacer), Date.now());
  }

  async load(namespace: string, key: string): Promise<unknown | null> {
    const row = this.db
      .prepare(`SELECT data FROM kv_store WHERE namespace = ? AND key = ?`)
      .get(namespace, key) as { data: string } | undefined;
    if (!row) return null;
    return JSON.parse(row.data, mapReviver) as unknown;
  }

  async list(namespace: string): Promise<string[]> {
    const rows = this.db
      .prepare(`SELECT key FROM kv_store WHERE namespace = ? ORDER BY key`)
      .all(namespace) as { key: string }[];
    return rows.map((r) => r.key);
  }

  async delete(namespace: string, key: string): Promise<boolean> {
    const result = this.db
      .prepare(`DELETE FROM kv_store WHERE namespace = ? AND key = ?`)
      .run(namespace, key);
    return result.changes > 0;
  }

  async exists(namespace: string, key: string): Promise<boolean> {
    const row = this.db
      .prepare(
        `SELECT 1 FROM kv_store WHERE namespace = ? AND key = ? LIMIT 1`,
      )
      .get(namespace, key);
    return row !== undefined;
  }

  /** Close the database connection */
  close(): void {
    this.db.close();
  }
}

// =============================================================================
// FACTORY
// =============================================================================

/** Create a persistence adapter of the given type */
export function createPersistenceAdapter(
  type: 'json' | 'sqlite',
  config: { baseDir?: string; dbPath?: string },
): PersistenceAdapter {
  if (type === 'json') {
    if (!config.baseDir) throw new Error('baseDir is required for json adapter');
    return new JSONFilePersistenceAdapter(config.baseDir);
  }
  if (!config.dbPath) throw new Error('dbPath is required for sqlite adapter');
  return new SQLitePersistenceAdapter(config.dbPath);
}

File: src/attocode/shared/persistence.py
"""Persistence adapters for shared state.

Provides file-based and SQLite-based backends for saving/loading
shared state across sessions and checkpoints.
"""

from __future__ import annotations

import json
import sqlite3
from abc import ABC, abstractmethod
from pathlib import Path
from typing import Any


class PersistenceAdapter(ABC):
    """Abstract persistence interface for shared state."""

    @abstractmethod
    async def save(self, namespace: str, key: str, data: Any) -> None:
        """Save data under a namespace/key pair."""

    @abstractmethod
    async def load(self, namespace: str, key: str) -> Any | None:
        """Load data by namespace/key, returning None if not found."""

    @abstractmethod
    async def list_keys(self, namespace: str) -> list[str]:
        """List all keys in a namespace."""

    @abstractmethod
    async def delete(self, namespace: str, key: str) -> bool:
        """Delete a key. Returns True if it existed."""

    @abstractmethod
    async def exists(self, namespace: str, key: str) -> bool:
        """Check if a key exists."""


class JSONFilePersistenceAdapter(PersistenceAdapter):
    """File-based persistence using JSON files.

    namespace = directory, key = JSON file.
    """

    def __init__(self, base_dir: str | Path) -> None:
        self.base_dir = Path(base_dir)
        self.base_dir.mkdir(parents=True, exist_ok=True)

    def _path(self, namespace: str, key: str) -> Path:
        ns_dir = self.base_dir / namespace
        ns_dir.mkdir(parents=True, exist_ok=True)
        return ns_dir / f"{key}.json"

    async def save(self, namespace: str, key: str, data: Any) -> None:
        path = self._path(namespace, key)
        path.write_text(json.dumps(data, default=str), encoding="utf-8")

    async def load(self, namespace: str, key: str) -> Any | None:
        path = self._path(namespace, key)
        if not path.exists():
            return None
        try:
            return json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            return None

    async def list_keys(self, namespace: str) -> list[str]:
        ns_dir = self.base_dir / namespace
        if not ns_dir.exists():
            return []
        return [p.stem for p in ns_dir.glob("*.json")]

    async def delete(self, namespace: str, key: str) -> bool:
        path = self._path(namespace, key)
        if path.exists():
            path.unlink()
            return True
        return False

    async def exists(self, namespace: str, key: str) -> bool:
        return self._path(namespace, key).exists()


class SQLitePersistenceAdapter(PersistenceAdapter):
    """SQLite-based persistence with namespace partitioning."""

    def __init__(self, db_path: str | Path) -> None:
        self.db_path = str(db_path)
        self._ensure_table()

    def _ensure_table(self) -> None:
        with sqlite3.connect(self.db_path) as conn:
            conn.execute("PRAGMA journal_mode=WAL")
            conn.execute("""
                CREATE TABLE IF NOT EXISTS shared_state (
                    namespace TEXT NOT NULL,
                    key TEXT NOT NULL,
                    data TEXT NOT NULL,
                    updated_at REAL NOT NULL DEFAULT (julianday('now')),
                    PRIMARY KEY (namespace, key)
                )
            """)

    async def save(self, namespace: str, key: str, data: Any) -> None:
        encoded = json.dumps(data, default=str)
        with sqlite3.connect(self.db_path) as conn:
            conn.execute(
                "INSERT OR REPLACE INTO shared_state (namespace, key, data) VALUES (?, ?, ?)",
                (namespace, key, encoded),
            )

    async def load(self, namespace: str, key: str) -> Any | None:
        with sqlite3.connect(self.db_path) as conn:
            row = conn.execute(
                "SELECT data FROM shared_state WHERE namespace = ? AND key = ?",
                (namespace, key),
            ).fetchone()
        if row is None:
            return None
        try:
            return json.loads(row[0])
        except json.JSONDecodeError:
            return None

    async def list_keys(self, namespace: str) -> list[str]:
        with sqlite3.connect(self.db_path) as conn:
            rows = conn.execute(
                "SELECT key FROM shared_state WHERE namespace = ?",
                (namespace,),
            ).fetchall()
        return [r[0] for r in rows]

    async def delete(self, namespace: str, key: str) -> bool:
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute(
                "DELETE FROM shared_state WHERE namespace = ? AND key = ?",
                (namespace, key),
            )
        return cursor.rowcount > 0

    async def exists(self, namespace: str, key: str) -> bool:
        with sqlite3.connect(self.db_path) as conn:
            row = conn.execute(
                "SELECT 1 FROM shared_state WHERE namespace = ? AND key = ? LIMIT 1",
                (namespace, key),
            ).fetchone()
        return row is not None
